File: Problems/0035. (E) Search Insert Position/Java/solution.java
class Solution {
  public int searchInsert(int[] nums, int target) {
    // binary search

    int lo = 0;
    int mid = 0;
    // nums is primitive array, use .length
    int hi = nums.length - 1;
    while (lo <= hi) {
      // no implicit type conversion on division
      mid = lo + (hi-lo) / 2;
      if (nums[mid] == target)
        return mid;
      else if (nums[mid] > target)
        hi = mid - 1;
      else
        lo = mid + 1;
    }
    return lo;
  }
}

File: Problems/0035. (E) Search Insert Position/Python3/solution.py
class Solution:
  def searchInsert(self, nums: List[int], target: int) -> int:
    # binary search

    lo, hi = 0, len(nums)-1
    while lo <= hi:
      mid = (hi-lo) // 2 + lo
      if nums[mid] == target:
        return mid
      elif nums[mid] < target:
        lo = mid+1
      else:
        hi = mid-1
    
    # want the insertion index if target is missing
    return lo

